sherpa/bin/streaming_conformer_rnnt/streaming_server.py

```diff
@@ -0,0 +1,632 @@
+
+        # Decrement so that it can accept new connections
+        self.current_active_connections -= 1
+
+        logging.info(f"Disconnected: {socket.remote_address}")
+
+    async def recv_audio_samples(
+        self,
+        socket: websockets.WebSocketServerProtocol,
+        last: Optional[bytes] = None,
+    ) -> Tuple[Optional[torch.Tensor], Optional[bytes]]:
+        """Receives a tensor from the client.
+
+        The message from the client contains two parts: header and payload
+
+            - the header contains 8 bytes in little endian format, specifying
+              the number of bytes in the payload.
+
+            - the payload contains either a binary representation of the 1-D
+              torch.float32 tensor or the bytes object b"Done" which means
+              the end of utterance.
+
+        Args:
+          socket:
+            The socket for communicating with the client.
+          last:
+            Previous received content.
+        Returns:
+          Return a tuple containing:
+            - A 1-D torch.float32 tensor containing the audio samples
+            - Data for the next chunk, if any
+         or return a tuple (None, None) meaning the end of utterance.
+        """
+        header_len = 8
+
+        if last is None:
+            last = b""
+
+        async def receive_header():
+            buf = last
+            async for message in socket:
+                buf += message
+                if len(buf) >= header_len:
+                    break
+            if buf:
+                header = buf[:header_len]
+                remaining = buf[header_len:]
+            else:
+                header = None
+                remaining = None
+
+            return header, remaining
+
+        header, received = await receive_header()
+
+        if header is None:
+            return None, None
+
+        expected_num_bytes = int.from_bytes(header, "little", signed=True)
+
+        async for message in socket:
+            received += message
+            if len(received) >= expected_num_bytes:
+                break
+
+        if not received or received == b"Done":
+            return None, None
+
+        this_chunk = received[:expected_num_bytes]
+        next_chunk = received[expected_num_bytes:]
+
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            # PyTorch warns that the underlying buffer is not writable.
+            # We ignore it here as we are not going to write it anyway.
+            if hasattr(torch, "frombuffer"):
+                # Note: torch.frombuffer is available only in torch>= 1.10
+                return (
+                    torch.frombuffer(this_chunk, dtype=torch.float32),
+                    next_chunk,
+                )  # noqa
+            else:
+                array = np.frombuffer(this_chunk, dtype=np.float32)
+                return torch.from_numpy(array), next_chunk
+
+
+@torch.no_grad()
+def main():
```

Comment:
Since `run_model_and_do_greedy_search()` is run in a separate thread, it looks like we also need to use `@torch.no_grad()` for it.